File: autostart.py
"""Manage "start with Windows" for Undertone via the HKCU Run registry key.

Uses only the stdlib ``winreg`` module. Enabling adds a value under the
current user's Run key so the app launches at login; disabling removes it.
"""

import sys
import winreg
from pathlib import Path

RUN_KEY = r"Software\Microsoft\Windows\CurrentVersion\Run"
VALUE_NAME = "Undertone"
LEGACY_VALUE_NAME = "PushToTalkSTT"


def _command() -> str:
    """Build the login launch command: the frozen exe, or pythonw + main.py."""
    if getattr(sys, "frozen", False):
        return f'"{sys.executable}"'
    pyw = Path(sys.executable).with_name("pythonw.exe")
    if not pyw.exists():
        pyw = Path(sys.executable)
    main_py = Path(__file__).resolve().parent / "main.py"
    return f'"{pyw}" "{main_py}"'


def migrate_legacy() -> None:
    """Move an old PushToTalkSTT Run entry to the Undertone name.

    Rewrites the command too, so the entry also heals if the project
    folder moved along with the rename.
    """
    try:
        with winreg.OpenKey(
            winreg.HKEY_CURRENT_USER, RUN_KEY, 0,
            winreg.KEY_QUERY_VALUE | winreg.KEY_SET_VALUE,
        ) as key:
            winreg.QueryValueEx(key, LEGACY_VALUE_NAME)
            winreg.DeleteValue(key, LEGACY_VALUE_NAME)
            winreg.SetValueEx(key, VALUE_NAME, 0, winreg.REG_SZ, _command())
    except (FileNotFoundError, OSError):
        pass


def is_enabled() -> bool:
    """Return True if the Run value exists (i.e. the user enabled autostart).

    Existence alone is the signal; a stale path still means the user opted in,
    and comparing against the current command would flap across relocations.
    """
    try:
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, RUN_KEY) as key:
            winreg.QueryValueEx(key, VALUE_NAME)
        return True
    except FileNotFoundError:
        return False
    except OSError:
        return False


def set_enabled(enabled: bool) -> None:
    """Enable or disable autostart by writing/removing the Run value."""
    if enabled:
        with winreg.CreateKey(winreg.HKEY_CURRENT_USER, RUN_KEY) as key:
            winreg.SetValueEx(key, VALUE_NAME, 0, winreg.REG_SZ, _command())
    else:
        try:
            with winreg.OpenKey(
                winreg.HKEY_CURRENT_USER, RUN_KEY, 0, winreg.KEY_SET_VALUE
            ) as key:
                winreg.DeleteValue(key, VALUE_NAME)
        except (FileNotFoundError, OSError):
            # Nothing to remove; disabling a missing value is a no-op.
            pass

File: electron/src/main/autostart.ts
import { execFile } from "node:child_process";

const RUN_KEY = String.raw`HKCU\Software\Microsoft\Windows\CurrentVersion\Run`;
const VALUE_NAME = "Undertone";
const LEGACY_VALUE_NAME = "PushToTalkSTT";

export interface RegistryRunner {
  run(arguments_: readonly string[]): Promise<number>;
}

export class AutostartManager {
  constructor(
    private readonly executable: string,
    private readonly runner: RegistryRunner = new RegExeRunner(),
  ) {}

  async reconcile(): Promise<void> {
    const [current, legacy] = await Promise.all([
      this.valueExists(VALUE_NAME),
      this.valueExists(LEGACY_VALUE_NAME),
    ]);
    if (legacy) await this.deleteValue(LEGACY_VALUE_NAME);
    if (current || legacy) await this.setEnabled(true);
  }

  async isEnabled(): Promise<boolean> {
    return await this.valueExists(VALUE_NAME);
  }

  async setEnabled(enabled: boolean): Promise<void> {
    if (!enabled) {
      await this.deleteValue(VALUE_NAME);
      await this.deleteValue(LEGACY_VALUE_NAME);
      return;
    }
    const command = `"${this.executable}" --autostart`;
    const code = await this.runner.run([
      "add", RUN_KEY, "/v", VALUE_NAME, "/t", "REG_SZ", "/d", command, "/f",
    ]);
    if (code !== 0) throw new Error("Could not enable Start with Windows");
  }

  private async valueExists(name: string): Promise<boolean> {
    return await this.runner.run(["query", RUN_KEY, "/v", name]) === 0;
  }

  private async deleteValue(name: string): Promise<void> {
    const code = await this.runner.run(["delete", RUN_KEY, "/v", name, "/f"]);
    if (code !== 0 && await this.valueExists(name)) {
      throw new Error("Could not update Start with Windows");
    }
  }
}

class RegExeRunner implements RegistryRunner {
  async run(arguments_: readonly string[]): Promise<number> {
    return await new Promise<number>((resolve, reject) => {
      execFile("reg.exe", [...arguments_], { windowsHide: true }, (error) => {
        if (error === null) {
          resolve(0);
          return;
        }
        if (typeof error.code === "number") resolve(error.code);
        else reject(error);
      });
    });
  }
}
